File: articles/grit-and-conscientiousness-overlap-distinction-and-debate/r/grit-conscientiousness-model.R
# Grit and Conscientiousness: Overlap, Distinction, and Debate
# Synthetic-data workflow for research-method demonstration only.
# Do not use this workflow to evaluate, rank, hire, discipline, or assess real people.

set.seed(42)

n <- 900

industriousness <- rnorm(n)
orderliness <- rnorm(n)
dependability <- rnorm(n)
responsibility <- rnorm(n)
achievement_striving <- rnorm(n)

conscientiousness <- (
  0.30 * industriousness +
  0.18 * orderliness +
  0.18 * dependability +
  0.17 * responsibility +
  0.17 * achievement_striving
)

perseverance_effort <- (
  0.55 * industriousness +
  0.25 * achievement_striving +
  rnorm(n, sd = 0.85)
)

consistency_interests <- (
  0.20 * achievement_striving +
  rnorm(n, sd = 1.00)
)

grit <- 0.60 * perseverance_effort + 0.40 * consistency_interests

prior_achievement <- rnorm(n)
social_support <- rnorm(n)
burnout <- rnorm(n)

long_term_progress <- (
  0.24 * conscientiousness +
  0.18 * grit +
  0.22 * prior_achievement +
  0.18 * social_support -
  0.20 * burnout +
  rnorm(n)
)

df <- data.frame(
  industriousness,
  orderliness,
  dependability,
  responsibility,
  achievement_striving,
  conscientiousness,
  perseverance_effort,
  consistency_interests,
  grit,
  prior_achievement,
  social_support,
  burnout,
  long_term_progress
)

c_median <- median(df$conscientiousness)
g_median <- median(df$grit)

df$profile <- ifelse(
  df$conscientiousness >= c_median & df$grit >= g_median,
  "high_conscientiousness_high_grit",
  ifelse(
    df$conscientiousness >= c_median & df$grit < g_median,
    "high_conscientiousness_low_grit",
    ifelse(
      df$conscientiousness < c_median & df$grit >= g_median,
      "low_conscientiousness_high_grit",
      "low_conscientiousness_low_grit"
    )
  )
)

dir.create("data/processed", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)

write.csv(df, "data/processed/grit-conscientiousness-modeled-data-r.csv", row.names = FALSE)

correlations <- cor(df[, c(
  "conscientiousness",
  "grit",
  "perseverance_effort",
  "consistency_interests",
  "industriousness",
  "achievement_striving",
  "long_term_progress"
)])

write.csv(correlations, "outputs/tables/grit-conscientiousness-correlations-r.csv")

profile_summary <- aggregate(
  cbind(
    long_term_progress,
    conscientiousness,
    grit,
    perseverance_effort,
    consistency_interests,
    burnout,
    social_support
  ) ~ profile,
  data = df,
  FUN = mean
)

write.csv(profile_summary, "outputs/tables/grit-conscientiousness-profile-summary-r.csv", row.names = FALSE)

model_conscientiousness <- lm(long_term_progress ~ conscientiousness, data = df)
model_grit <- lm(long_term_progress ~ grit, data = df)
model_combined <- lm(long_term_progress ~ conscientiousness + grit, data = df)

model_facets <- lm(
  long_term_progress ~ conscientiousness +
    perseverance_effort + consistency_interests,
  data = df
)

model_contextual <- lm(
  long_term_progress ~ conscientiousness +
    perseverance_effort + consistency_interests +
    prior_achievement + social_support + burnout,
  data = df
)

comparison <- data.frame(
  model = c(
    "conscientiousness_only",
    "grit_only",
    "conscientiousness_plus_grit",
    "conscientiousness_plus_grit_facets",
    "contextual_model"
  ),
  r_squared = c(
    summary(model_conscientiousness)$r.squared,
    summary(model_grit)$r.squared,
    summary(model_combined)$r.squared,
    summary(model_facets)$r.squared,
    summary(model_contextual)$r.squared
  ),
  adjusted_r_squared = c(
    summary(model_conscientiousness)$adj.r.squared,
    summary(model_grit)$adj.r.squared,
    summary(model_combined)$adj.r.squared,
    summary(model_facets)$adj.r.squared,
    summary(model_contextual)$adj.r.squared
  )
)

write.csv(comparison, "outputs/tables/grit-conscientiousness-model-comparison-r.csv", row.names = FALSE)

print(round(correlations, 3))
print(round(profile_summary, 3))
print(round(comparison, 4))

File: articles/grit-and-conscientiousness-overlap-distinction-and-debate/python/grit_conscientiousness_model.py
"""
Grit and Conscientiousness: Overlap, Distinction, and Debate

Synthetic-data workflow for modeling grit and conscientiousness as overlapping
but distinguishable predictors, with incremental-validity comparisons.

This script is for article support and research-method demonstration only.
It should not be used to assess, rank, hire, discipline, or evaluate real people.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-conscientiousness-modeled-data-python.csv"
TABLE_OUT = ROOT / "outputs" / "tables" / "grit-conscientiousness-model-comparison-python.csv"
PROFILE_OUT = ROOT / "outputs" / "tables" / "grit-conscientiousness-profile-summary-python.csv"
CORR_OUT = ROOT / "outputs" / "tables" / "grit-conscientiousness-correlations-python.csv"

rng = np.random.default_rng(42)
n = 900

industriousness = rng.normal(0, 1, n)
orderliness = rng.normal(0, 1, n)
dependability = rng.normal(0, 1, n)
responsibility = rng.normal(0, 1, n)
achievement_striving = rng.normal(0, 1, n)

conscientiousness = (
    0.30 * industriousness
    + 0.18 * orderliness
    + 0.18 * dependability
    + 0.17 * responsibility
    + 0.17 * achievement_striving
)

perseverance_effort = (
    0.55 * industriousness
    + 0.25 * achievement_striving
    + rng.normal(0, 0.85, n)
)

consistency_interests = (
    0.20 * achievement_striving
    + rng.normal(0, 1.00, n)
)

grit = 0.60 * perseverance_effort + 0.40 * consistency_interests

prior_achievement = rng.normal(0, 1, n)
social_support = rng.normal(0, 1, n)
burnout = rng.normal(0, 1, n)

long_term_progress = (
    0.24 * conscientiousness
    + 0.18 * grit
    + 0.22 * prior_achievement
    + 0.18 * social_support
    - 0.20 * burnout
    + rng.normal(0, 1, n)
)

df = pd.DataFrame(
    {
        "industriousness": industriousness,
        "orderliness": orderliness,
        "dependability": dependability,
        "responsibility": responsibility,
        "achievement_striving": achievement_striving,
        "conscientiousness": conscientiousness,
        "perseverance_effort": perseverance_effort,
        "consistency_interests": consistency_interests,
        "grit": grit,
        "prior_achievement": prior_achievement,
        "social_support": social_support,
        "burnout": burnout,
        "long_term_progress": long_term_progress,
    }
)

c_median = df["conscientiousness"].median()
g_median = df["grit"].median()

conditions = [
    (df["conscientiousness"] >= c_median) & (df["grit"] >= g_median),
    (df["conscientiousness"] >= c_median) & (df["grit"] < g_median),
    (df["conscientiousness"] < c_median) & (df["grit"] >= g_median),
    (df["conscientiousness"] < c_median) & (df["grit"] < g_median),
]
choices = [
    "high_conscientiousness_high_grit",
    "high_conscientiousness_low_grit",
    "low_conscientiousness_high_grit",
    "low_conscientiousness_low_grit",
]
df["profile"] = np.select(conditions, choices, default="unclassified")

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
TABLE_OUT.parent.mkdir(parents=True, exist_ok=True)
PROFILE_OUT.parent.mkdir(parents=True, exist_ok=True)
CORR_OUT.parent.mkdir(parents=True, exist_ok=True)

df.to_csv(DATA_OUT, index=False)

correlations = df[
    [
        "conscientiousness",
        "grit",
        "perseverance_effort",
        "consistency_interests",
        "industriousness",
        "achievement_striving",
        "long_term_progress",
    ]
].corr()

correlations.to_csv(CORR_OUT)

model_conscientiousness = sm.OLS(
    df["long_term_progress"],
    sm.add_constant(df[["conscientiousness"]]),
).fit()

model_grit = sm.OLS(
    df["long_term_progress"],
    sm.add_constant(df[["grit"]]),
).fit()

model_combined = sm.OLS(
    df["long_term_progress"],
    sm.add_constant(df[["conscientiousness", "grit"]]),
).fit()

model_facets = sm.OLS(
    df["long_term_progress"],
    sm.add_constant(
        df[
            [
                "conscientiousness",
                "perseverance_effort",
                "consistency_interests",
            ]
        ]
    ),
).fit()

model_contextual = sm.OLS(
    df["long_term_progress"],
    sm.add_constant(
        df[
            [
                "conscientiousness",
                "perseverance_effort",
                "consistency_interests",
                "prior_achievement",
                "social_support",
                "burnout",
            ]
        ]
    ),
).fit()

comparison = pd.DataFrame(
    {
        "model": [
            "conscientiousness_only",
            "grit_only",
            "conscientiousness_plus_grit",
            "conscientiousness_plus_grit_facets",
            "contextual_model",
        ],
        "r_squared": [
            model_conscientiousness.rsquared,
            model_grit.rsquared,
            model_combined.rsquared,
            model_facets.rsquared,
            model_contextual.rsquared,
        ],
        "adjusted_r_squared": [
            model_conscientiousness.rsquared_adj,
            model_grit.rsquared_adj,
            model_combined.rsquared_adj,
            model_facets.rsquared_adj,
            model_contextual.rsquared_adj,
        ],
    }
)

profile_summary = (
    df.groupby("profile", as_index=False)
    .agg(
        n=("profile", "size"),
        mean_conscientiousness=("conscientiousness", "mean"),
        mean_grit=("grit", "mean"),
        mean_perseverance_effort=("perseverance_effort", "mean"),
        mean_consistency_interests=("consistency_interests", "mean"),
        mean_long_term_progress=("long_term_progress", "mean"),
        mean_burnout=("burnout", "mean"),
        mean_social_support=("social_support", "mean"),
    )
)

comparison.to_csv(TABLE_OUT, index=False)
profile_summary.to_csv(PROFILE_OUT, index=False)

print("Saved modeled dataset:", DATA_OUT)
print("Saved correlations:", CORR_OUT)
print("Saved model comparison:", TABLE_OUT)
print("Saved profile summary:", PROFILE_OUT)
print(comparison.round(4))
print(profile_summary.round(4))
